### ISSUE TITLE
How do you destroy the editor object?

### ISSUE THREAD

**Issue by ariera** [quilljs/quill#138]
Is there a way to programatically stop using the editor, so it removes the bindings, and frees memory without creating leaks?

Something like:

``` javascript
editor = new Quill('#text')
editor.remove()
```

I tried with `delete editor` but lots of errors start appearing on the console whenever I try to do anything with it.

**Comment by MarcGodard** [comment]
Does this still exist? I am getting "destroy is not a function"

**Comment by gieoon** [comment]
It's a workaround, but I manually removed the children by targeting their shared parent:
```
editorParentElement.querySelector('.ql-toolbar').style.display = "none";
editorParentElement.querySelector('.ql-container').style.display = "none";
```

**Comment by canadaduane** [comment]
> Does this still exist? I am getting "destroy is not a function"

See https://github.com/quilljs/quill/blob/9cf0285caa6514356a7bba9db132ec7229eb254c/docs/guides/upgrading-to-1-0.md

> `destroy` removed - No longer necessary as Quill no longer keeps references to editor instances, allowing Javascript garbage collectors to work as expected.

**Comment by lesichkovm** [comment]
This is a regression. Its definitely needed

**Comment by drkrtj** [comment]
Has anybody come up with a way to destroy the instance yet? I am receiving "quill Overwriting modules/counter".

**Comment by gbrits** [comment]
This worked for me:

```
Quill.prototype.destroy = function () {
  // Disable the editor to prevent further user input
  this.enable(false);

  // Remove event listeners managed by Quill
  this.emitter.listeners = {};
  this.emitter.off();

  // Clear clipboard event handlers
  if (this.clipboard && this.clipboard.off) {
    this.clipboard.off();
  }

  // Remove keyboard bindings
  this.keyboard.bindings = {};

  // Clear history stack
  this.history.clear();

  // Remove toolbar event handlers (if toolbar module exists)
  if (this.theme && this.theme.modules.toolbar) {
    this.theme.modules.toolbar.container.remove();
  }

  // Remove tooltip (if present)
  if (this.theme && this.theme.tooltip) {
    this.theme.tooltip.root.remove();
  }

  // Remove all Quill-added classes from the container
  const container = this.container;
  container.classList.forEach((cls) => {
    if (cls.startsWith('ql-')) {
      container.classList.remove(cls);
    }
  });

  // Restore the original container content (before Quill modified it)
  container.innerHTML = this.root.innerHTML;

  // Remove Quill-specific DOM elements
  this.root.remove();

  // Nullify references to allow garbage collection
  this.root = null;
  this.scroll = null;
  this.emitter = null;
  this.clipboard = null;
  this.keyboard = null;
  this.history = null;
  this.theme = null;
  this.container = null;

  // Override isEnabled to prevent errors after destruction
  this.isEnabled = function() {
    return false;
  };

  // Remove the instance from Quill's internal registry (if any)
  if (Quill.instances && Quill.instances[this.id]) {
    delete Quill.instances[this.id];
  }
};
```

Just add it into your bootstrap.js or your app.js where you initialise Quill to appropriately add the prototype method.
In my case I initialised quill into a variable like:

```
if(!window.noteQuill) {
window.noteQuill = new Quill...
...
```

So when I killed it, I ran the new method and also nullified the container:

```
window.noteQuill.destory(); window.noteQuill = null;
```

Hopefully helps you guys out.

**Comment by drkrtj** [comment]
Thank you very much. 


     On Sunday, May 18, 2025 at 12:12:32 AM PDT, Grant ***@***.***> wrote:  
 
 gbrits left a comment (slab/quill#138)
This worked for me:
Quill.prototype.destroy = function () {  // Disable the editor to prevent further user input  this.enable(false);  // Remove event listeners managed by Quill  this.emitter.listeners = {};  this.emitter.off();  // Clear clipboard event handlers  if (this.clipboard && this.clipboard.off) {    this.clipboard.off();  }  // Remove keyboard bindings  this.keyboard.bindings = {};  // Clear history stack  this.history.clear();  // Remove toolbar event handlers (if toolbar module exists)  if (this.theme && this.theme.modules.toolbar) {    this.theme.modules.toolbar.container.remove();  }  // Remove tooltip (if present)  if (this.theme && this.theme.tooltip) {    this.theme.tooltip.root.remove();  }  // Remove all Quill-added classes from the container  const container = this.container;  container.classList.forEach((cls) => {    if (cls.startsWith('ql-')) {      container.classList.remove(cls);    }  });  // Restore the original container content (before Quill modified it)  container.innerHTML = this.root.innerHTML;  // Remove Quill-specific DOM elements  this.root.remove();  // Nullify references to allow garbage collection  this.root = null;  this.scroll = null;  this.emitter = null;  this.clipboard = null;  this.keyboard = null;  this.history = null;  this.theme = null;  this.container = null;  // Override isEnabled to prevent errors after destruction  this.isEnabled = function() {    return false;  };  // Remove the instance from Quill's internal registry (if any)  if (Quill.instances && Quill.instances[this.id]) {    delete Quill.instances[this.id];  }};
Just add it into your bootstrap.js or your app.js where you initialise Quill to appropriately add the prototype method.
In my case I initialised quill into a variable like:
if(!window.noteQuill) {window.noteQuill = new Quill......
So when I killed it, I ran the new method and also nullified the container:
window.noteQuill.destory(); window.noteQuill = null;
Hopefully helps you guys out.
